urllib3/connection.py

```diff
@@ -265,7 +265,14 @@ def connect(self):
                     'for details.)'.format(hostname)),
                     SubjectAltNameWarning
                 )
-            match_hostname(cert, self.assert_hostname or hostname)
+
+            # In case the hostname is an IPv6 address, strip the square
+            # brackets from it before using it to validate. Non IPv6 addresses
+            # should not start/end with square brackets, so this should be
+            # safe.
```

Comment:
It's not that it demands that, it's that we strip `[]` from IPv6 hostnames (because of httplib on some Python 3 version (I think 3.2)) and when we then pass them to `ssl_match_hostname` we have a mismatch between `::1` and `[::1]`.